tools/src/scie_pants/pants_version.py

```diff
@@ -162,5 +182,120 @@ def configure_version():
         pants_config.write_text(tomlkit.dumps(config))
 
     return configure_version, determine_tag_version(
-        ptex, pants_version, find_links_dir, github_api_bearer_token
+        ptex, pants_version, find_links_dir, github_api_bearer_token, bootstrap_urls_path
     )
+
+
+PYTHON_IDS = {
+    # N.B.: These values must match the lift TOML interpreter ids.
+    # Important: all pythons used in pants_python_versions.json must be represented in this list.
+    "cp38": "cpython38",
+    "cp39": "cpython39",
+    "cp310": "cpython310",
+    "cp311": "cpython311",
+}
+
+
+def determine_pex_url_and_python_id(
+    ptex: Ptex,
+    version: Version,
+    bootstrap_urls_path: str | None,
+) -> tuple[str, str]:
+    uname = os.uname()
+    platform = f"{uname.sysname.lower()}_{uname.machine.lower()}"
+    res = get_pex_url_and_python_id(ptex, version, platform, bootstrap_urls_path)
+    if not res:
+        bootstrap_urls_info = (
+            (
+                f"It may also be a missing Pants PEX entry in the PANTS_BOOTSTRAP_URLS file {bootstrap_urls_path} for "
+                f"a name like `pants.{version}-cpNN-{platform}.pex`.\n\n"
+            )
+            if bootstrap_urls_path
+            else ""
+        )
+        fatal(
+            f"Unknown Pants release: {version}.\n\n"
+            "If this is unexpected (you are using a known good Pants version), try upgrading scie-pants first.\n"
+            f"It may also be that the platform {platform} isn't supported for this version of Pants.\n"
+            + bootstrap_urls_info
+            + "To get help, please visit: https://www.pantsbuild.org/community/getting-help"
+        )
+    pex_url, python = res
+    if python not in PYTHON_IDS:
+        # Should not happen... but if we mess up, this is a nicer error message rather than blowing up.
+        fatal(f"This version of scie-pants does not support {python!r}.")
+    return pex_url, PYTHON_IDS[python]
+
+
+def get_pex_url_and_python_id(
+    ptex: Ptex,
+    version: Version,
+    platform: str,
+    bootstrap_urls_path: str | None,
+) -> tuple[str, str] | None:
+    if bootstrap_urls_path:
+        bootstrap_urls = json.loads(Path(bootstrap_urls_path).read_text())
+        ptex_urls = bootstrap_urls.get("ptex")
+        if ptex_urls is None:
+            raise ValueError(
+                f"Missing 'ptex' key in PANTS_BOOTSTRAP_URLS file: {bootstrap_urls_path}"
+            )
+    else:
+        ptex_urls = None
+
+    py = get_python_id_for_pants_version(version)
+    if py:
+        pex_url = get_download_url(version, platform, py, ptex_urls)
```

Comment:
Do we know if these URLs have rate limits for `HEAD` requests?

Reply:
Unsure. It seems there could be a limit approx. 5k req/ip-address/hour.
https://stackoverflow.com/a/74960542/444060